contracts/strategies/traderjoe/TraderJoe.sol

```diff
@@ -245,41 +219,335 @@ contract TraderJoe is UUPSUpgradeable, StrategyOwnablePausableBaseUpgradeable {
+        // solhint-disable-next-line reason-string
+        require(
+            binsAmount == allocationsAmount,
+            "TraderJoe: the number of bin IDs and allocations are different"
+        );
+
+        // solhint-disable-next-line reason-string
+        require(
+            binsAmount >= 1 && binsAmount <= 51,
+            "TraderJoe: too few or too many bin ID"
+        );
+        // No need to check the number of allocations.
+
+        for (uint256 i; i < binsAmount; i++) {
+            require(binIds[i] <= type(uint24).max, "TraderJoe: too big bin ID");
+        }
+
+        uint256 allocations;
+
+        for (uint256 i; i < binsAmount; i++) {
+            allocations += binAllocations[i];
+        }
+
+        // solhint-disable-next-line reason-string
+        require(
+            allocations == 1e3,
+            "TraderJoe: the sum of allocations must be 1e3"
+        );
+    }
+
+    function __initialize(
+        ITraderJoeLBPair lbPair,
+        ITraderJoeLBRouter lbRouter,
+        uint256 binStep,
+        uint256[] calldata binIds,
+        uint256[] calldata binAllocations
+    ) private {
+        TraderJoeStorage storage strategyStorage = TraderJoeStorageLib
+            .getStorage();
+
+        strategyStorage.lbPair = lbPair;
+        strategyStorage.lbRouter = lbRouter;
+        strategyStorage.tokenX = IERC20Upgradeable(lbPair.tokenX());
+        strategyStorage.tokenY = IERC20Upgradeable(lbPair.tokenY());
+        strategyStorage.binStep = binStep;
+        strategyStorage.binIds = binIds;
+        strategyStorage.binAllocations = binAllocations;
+
+        if (strategyStorage.tokenX == depositToken) {
+            strategyStorage.pairDepositToken = IERC20Upgradeable(
+                strategyStorage.tokenY
+            );
+        } else if (strategyStorage.tokenY == depositToken) {
+            strategyStorage.pairDepositToken = IERC20Upgradeable(
+                strategyStorage.tokenX
+            );
+        } else {
+            revert InvalidTraderJoeLBPair();
+        }
+    }
+
+    function __deposit(
+        uint256 depositTokenAmount,
+        uint256 pairDepositTokenAmount
+    ) private {
+        TraderJoeStorage storage strategyStorage = TraderJoeStorageLib
+            .getStorage();
+
+        uint256 pairDepositTokenBefore = strategyStorage
+            .pairDepositToken
+            .balanceOf(address(this)) - pairDepositTokenAmount;
+
+        uint256 amountXIn;
+        uint256 amountYIn;
+
+        if (strategyStorage.tokenX == depositToken) {
+            amountXIn = depositTokenAmount;
+            amountYIn = pairDepositTokenAmount;
+        } else {
+            amountXIn = pairDepositTokenAmount;
+            amountYIn = depositTokenAmount;
+        }
+
+        uint256 binsAmount = strategyStorage.binIds.length;
+
+        (, , uint256 activeId) = strategyStorage.lbPair.getReservesAndId();
+
+        uint256 totalAmount = amountXIn + amountYIn;
```

Comment:
Can we please make a comment here regarding assumptions: saying that we assume AssetX and AssetY worth the same amount in term of USDC

Reply:
Little correction, what we assume is their price are the same (or nearly the same). Their price in USDC can be different and it doesn't matter. Here is a special case in that asset Y is USDC.

Reply:
right, just please document it here: if we are going to do some generalization later, we can just 'grep' for all the assumptions